java-checks/src/main/java/org/sonar/java/checks/InvalidDateValuesCheck.java

```diff
@@ -0,0 +1,244 @@
+import org.sonar.java.checks.methods.MethodInvocationMatcher;
+import org.sonar.java.resolve.Symbol;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.UnaryExpressionTree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import javax.annotation.Nullable;
+import java.text.MessageFormat;
+import java.util.List;
+import java.util.Map;
+
+@Rule(
+  key = "S2110",
+  name = "Invalid \"Date\" values should not be used",
+  tags = {"bug"},
+  priority = Priority.CRITICAL)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.INSTRUCTION_RELIABILITY)
+@SqaleConstantRemediation("5min")
+@ActivatedByDefault
+public class InvalidDateValuesCheck extends AbstractMethodDetection {
+
+  private static final String[] GREGORIAN_PARAMETERS = {"year", "month", "dayOfMonth", "hourOfDay", "minute", "second"};
+
+  private static final List<MethodInvocationMatcher> DATE_METHODS_COMPARISON = ImmutableList.<MethodInvocationMatcher>builder()
+    .add(MethodInvocationMatcher.create().typeDefinition("java.util.Calendar").name("get").addParameter("int"))
+    .add(dateMethodInvocationMatcherGetter("java.util.Date", "getDate"))
+    .add(dateMethodInvocationMatcherGetter("java.util.Date", "getMonth"))
+    .add(dateMethodInvocationMatcherGetter("java.util.Date", "getHours"))
+    .add(dateMethodInvocationMatcherGetter("java.util.Date", "getMinutes"))
+    .add(dateMethodInvocationMatcherGetter("java.util.Date", "getSeconds"))
+    .add(dateMethodInvocationMatcherGetter("java.sql.Date", "getDate"))
+    .add(dateMethodInvocationMatcherGetter("java.sql.Date", "getMonth"))
+    .add(dateMethodInvocationMatcherGetter("java.sql.Date", "getHours"))
+    .add(dateMethodInvocationMatcherGetter("java.sql.Date", "getMinutes"))
+    .add(dateMethodInvocationMatcherGetter("java.sql.Date", "getSeconds"))
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.<Tree.Kind>builder().addAll(super.nodesToVisit())
+      .add(Tree.Kind.EQUAL_TO)
+      .add(Tree.Kind.NOT_EQUAL_TO)
+      .build();
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    super.visitNode(tree);
+    if (hasSemantic() && tree.is(Tree.Kind.EQUAL_TO, Tree.Kind.NOT_EQUAL_TO)) {
+      BinaryExpressionTree binaryExpressionTree = (BinaryExpressionTree) tree;
+      String name = getThresholdToCheck(binaryExpressionTree.leftOperand());
```

Comment:
Maybe call the variable `threshold` instead of `name`, which is a bit confusing.